src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -3,99 +3,296 @@
-            ISerializerService serializer,
-            SolutionReplicationContext context,
-            Checksum solutionChecksum,
-            ReadOnlyMemory<Checksum> checksums,
-            CancellationToken cancellationToken)
+        // Create a channel to communicate between the searching and writing tasks.  This allows the searching task to
+        // find items, add them to the channel synchronously, and immediately continue searching for more items.
+        // Concurrently, the writing task can read from the channel and write the items to the pipe-writer.
+        var channel = Channel.CreateUnbounded<(Checksum checksum, object asset)>(s_channelOptions);
+
+        // When cancellation happens, attempt to close the channel.  That will unblock the task writing the assets
+        // to the pipe.
+        using var _ = cancellationToken.Register(() => channel.Writer.TryComplete(new OperationCanceledException(cancellationToken)));
+
+        // Keep track of how many checksums we found.  We must find all the checksums we were asked to find.
+        var foundChecksumCount = 0;
+
+        // Spin up a task to go search for all the requested checksums, adding results to the channel.
+        var findAssetsTask = FindAllAssetsAsync();
+
+        // Spin up a task to read from the channel and write out the assets to the pipe-writer.
+        var writeAssetsTask = WriteAllAssetsToPipeAsync();
+
+        // Wait for both the searching and writing tasks to finish.
+        await Task.WhenAll(findAssetsTask, writeAssetsTask).ConfigureAwait(false);
+        cancellationToken.ThrowIfCancellationRequested();
+
+        // If we weren't canceled, we better have found and written out all the expected assets.
+        Contract.ThrowIfTrue(foundChecksumCount != checksums.Length);
+
+        return;
+
+        async Task FindAllAssetsAsync()
         {
-            using var writer = new ObjectWriter(stream, leaveOpen: true, cancellationToken);
+            await Task.Yield();
 
-            // This information is not actually needed on the receiving end.  However, we still send it so that the
-            // receiver can assert that both sides are talking about the same solution snapshot and no weird invariant
-            // breaks have occurred.
-            solutionChecksum.WriteTo(writer);
+            try
+            {
+                await scope.FindAssetsAsync(
+                    assetPath,
+                    checksums,
+                    (checksum, asset) => channel.Writer.TryWrite((checksum, asset)),
+                    cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                // If Something went wrong ensure that we complete the channel so that the writing task will stop.
+                // Also bubble the exception out so that the outer Task.WhenAll will bubble it up.
+                channel.Writer.TryComplete(ex);
```

Comment:
>channel.Writer.TryComplete(ex); [](http://example.com/codeflow?start=16&length=31)

Should this not happen for Cancellation as the Register method above handles that, or is it ok for TryComplete to occur twice in that scenario?

Reply:
Oh, TryComplete must be fine being called multiple times as you do it in the finally after catching the exception (hopefully it's ok being called 3 times in the cancellation case) :)

Reply:
Maybe move the TryComplete inside the try part of the block to avoid one of those calls

Reply:
it's fine for try-complete to happen any number of times (that's why it's "try").  this allows anyone to win in a case like thsi.  this codepath also handles non-cancellation exception.

Reply:
this also allows the channel to complete quickly, in the case where the finder may not notice cancellation for a moement.

Reply:
ok, the only nit I really have is that the finally block isn't really needed, the TryComplete in there could just be moved inside the try block or just moved outside the try/catch completely.

Reply:
i was on the fence about that... something made me feel better about the finally.  but you're right.  it should not be necessary.  i'll move.

Reply:
switched to a apttern we use eleswhere for this.